doc/proposals/scorecard-user-tests.md

```diff
@@ -0,0 +1,153 @@
+      status:
+        scorecard_function_length:
+          nodes: 3
+    modifications:
+      - spec:
+          size: 4
+        expected:
+          resources:
+            - kind: Deployment
+              name: "example_memcached"
+              status:
+                readyReplicas: 4
+          status:
+            scorecard_function_length:
+              nodes: 4
+```
+
+This is what the golang structs would look like, including comments describing each field:
+
+```go
+// Struct containing a user defined test. User passes tests as an array using the `functional_tests` viper config
+type UserDefinedTest struct {
+    // Path to cr to be used for testing
+    CRPath string `mapstructure:"cr"`
+    // Expected resources and status
+    Expected Expected `mapstructure:"expected"`
+    // Sub-tests modifying a few fields with expected changes
+    Modifications []Modification `mapstructure:"modifications"`
+}
+
+type Expected struct {
+	// Resources expected to be created after the operator reacts to the CR
+	Resources []map[string]interface{} `mapstructure:"resources"`
+	// Expected values in CR's status after the operator reacts to the CR
+	Status map[string]interface{} `mapstructure:"status"`
+}
+
+// Modifications specifies a spec field to change in the CR with the expected results
+type Modification struct {
+    // a map of the spec fields to modify
+    Spec map[string]interface{} `mapstructure:"spec"`
+    // Expected resources and status
+    Expected Expected `mapstructure:"expected"`
+}
+```
+
+For `Status` and `Resources` fields, we can implement a bit of extra computation instead of simple string checking. For instance,
+in the memcached-operator test, we should expect that the length of the `nodes` field (which is an array) has a certain length. To implement functions like
+these, we can create some functions for these checks that are prepended by `scorecard_function_` and take an array of objects. For instance, in the above
+example, `scorecard_function_length` would check that each field listed under it matches the specified length (like `nodes: 4`). If the yaml key does not
+start with `scorecard_function_`, we do a simple match (like `status/readyReplicas: 4`).
+
+This design would allow us to replace the old "Operator actions are reflected in status" (which would be tested by the `expected/status` check) and
+"Writing into CRs has an effect" (which would be tested by the `expected/resources` check) tests.
+
+### Plugin System
+
+In order to increase the flexibility of the user defined tests and allow users to implement more complex E2E style tests for scorecard,
+the user-defined tests will be implemented via a plugin system. Users will specify the path of the script they wish to run as well
+as environment variable to set for the command. The command would then print out the result as JSON to stdout. Here is an example:
+
+```yaml
+user_defined_tests:
+- path: "scorecard/simple-scorecard.sh"
```

Comment:
What would be an example of the script?

Reply:
Assuming we have a command called `simple-scorecard`, we could do:

```bash
#!/bin/bash

simple-scorecard --config $CONFIG_FILE
```

For the second example I have here, which would be a modified version of our test framework's tests to support the JSON output, we could do something like this:

```bash
#!/bin/bash

operator-sdk test local $TEST_DIR --enable-scorecard $ENABLE_SCORECARD --namespaced-manifest $NAMESPACED_MANIFEST --go-test-flags $GO_TEST_FLAGS
```

The scripts are intended to be just a simple wrappers, but the scripts could do more complex things based on the environment variables if we want them to

Reply:
Thanks!